primitives/core/src/offchain/storage.rs

```diff
@@ -58,3 +77,63 @@ impl OffchainStorage for InMemOffchainStorage {
 		}
 	}
 }
+
+
+
+
+/// Change to be applied to the offchain worker db in regards to a key.
+#[derive(Debug,Clone,Hash,Eq,PartialEq)]
+pub enum OffchainOverlayedChange {
+	/// Remove the data associated with the key
+	Remove,
+	/// Overwrite the value of an associated key
+	SetValue(Vec<u8>),
+}
+
+/// In-memory storage for offchain workers.
+#[derive(Debug, Clone)]
+pub struct OffchainOverlayedChanges {
+	storage: HashMap<Vec<u8>, OffchainOverlayedChange>,
+}
+
+/// Create an empty set of overlay changes.
+///
+/// The `Default` impl has to be done manually to avoid requirement of
+/// a `Default` impl for `OffchainOverlayedChange`
+impl Default for OffchainOverlayedChanges {
+	fn default() -> Self {
+		Self {
+			storage: HashMap::new()
+		}
+	}
+}
+
+
+impl OffchainOverlayedChanges {
+	/// Consume the offchain storage and iterate over all key value pairs.
+	pub fn into_iter(self) -> impl Iterator<Item=(Vec<u8>,OffchainOverlayedChange)> {
+		self.storage.into_iter()
+	}
+
+	/// Iterate over all key value pairs by reference.
+	pub fn iter<'a>(&'a self) -> impl Iterator<Item=(&'a Vec<u8>,&'a OffchainOverlayedChange)> {
+		self.storage.iter()
+	}
+
+	/// Drain all elements of changeset.
+	pub fn drain<'a,'d>(&'a mut self) -> impl Iterator<Item=(Vec<u8>, OffchainOverlayedChange)> + 'd where 'a : 'd {
+		self.storage.drain()
+	}
+
+	/// Remove a key and it's associated value from the offchain database.
+	pub fn remove(&mut self, prefix: &[u8], key: &[u8]) {
+		let key: Vec<u8> = prefix.iter().chain(key).cloned().collect();
+		let _ = self.storage.insert(key, OffchainOverlayedChange::Remove);
+	}
+
+	/// Set the value associated with a key under a prefix to the value provided.
+	pub fn set(&mut self, prefix: &[u8], key: &[u8], value: &[u8]) {
```

Comment:
```suggestion
	pub fn set(&mut self, prefix: &[u8], key: &[u8], value: Vec<u8>) {
```
Maybe be explicit with the API, perhaps we can save some copying this way.